networkx/algorithms/shortest_paths/weighted.py

```diff
@@ -984,3 +985,82 @@ def bidirectional_dijkstra(G, source, target, weight='weight'):
                         revpath.reverse()
                         finalpath = paths[0][w] + revpath[1:]
     raise nx.NetworkXNoPath("No path between %s and %s." % (source, target))
+
+
+def johnson(G, weight='weight', new_weight=None):
+    """Compute shortest paths between all nodes in a weighted graph using
+    Johnson's algorithm.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+
+    weight: string, optional (default='weight')
+        Edge data key corresponding to the edge weight.
+
+    new_weight: string, optional (default=None)
+        Edge data key corresponding to the new edge weight after graph transformation.
+
+    Returns
+    -------
+    distance : dictionary
+       Dictionary, keyed by source and target, of shortest paths.
+
+    Raises
+    ------
+    NetworkXError
+       If given graph is not weighted.
+
+    Examples
+    --------
+    >>> import networkx as nx
+    >>> graph = nx.DiGraph()
+    >>> graph.add_weighted_edges_from([('0', '3', 3), ('0', '1', -5), ('0', '2', 2), ('1', '2', 4),('2', '3', 1)])
+    >>> paths = nx.johnson(graph,weight='weight')
+    >>> paths['0']['2']
+    ['0', '1', '2']
+
+    Notes
+    ------
+    Johnson's algorithm is suitable even for graphs with negative weights. It
+    works by using the Bellman–Ford algorithm to compute a transformation of
+    the input graph that removes all negative weights, allowing Dijkstra's
+    algorithm to be used on the transformed graph.
+
+    It may be faster than Floyd - Warshall algorithm in sparse graphs.
+    Algorithm complexity: O(V^2 * logV + V * E)
+
+    See Also
+    --------
+    floyd_warshall_predecessor_and_distance
+    floyd_warshall_numpy
+    all_pairs_shortest_path
+    all_pairs_shortest_path_length
+    all_pairs_dijkstra_path
+    bellman_ford
+    """
+    new_node = nx.utils.generate_unique_node()
+    G.add_node(new_node)
+    G.add_weighted_edges_from((new_node, node, 0) for node in G)
+
+    # Calculate distance of shortest paths
+    dist = nx.bellman_ford(G, source=new_node, weight=weight)[1]
+
+    delete = False
+    if new_weight is None:
+        delete = True
+        new_weight = nx.utils.generate_unique_node()
```

Comment:
Copying the graph is unnecessary. In order to compute node potentials using Bellman--Ford you do not need to add a separate node. You just assign zero distance labels of all nodes and start the relaxation loop.

The following Dijkstra using reduced costs (`w[u][v] + p[u] - p[v]`) instead of raw weights (`w[u][v]`) warrants introducing a dedicated function for that purpose. It has use cases in other algorithm such as augmenting path-based minimum cost flow algorithms. (Maybe we can use unify this with #762.)

It is unnecessary restrictive to prevent users from relying on stability of iteration order (this is different from relying on some particular order being picked). Python dicts [https://docs.python.org/2/library/stdtypes.html#dict.items](guarantee) that the iteration order is stable when there is no intervening modifications. We should extend this benefit to users without requiring them to jump through hoops like changing to `OrderDict`s.